### ISSUE TITLE
Clarify definition of "smaller than a hemisphere"

### ISSUE THREAD

**Issue by urschrei** [d3/d3-geo#138]
In https://github.com/d3/d3-geo/pull/79#issuecomment-281130284, @mbostock clarified the required winding order of polygons as dependent upon their size (smaller than, or larger than a hemisphere). Could someone state this in more concrete terms? I.e. is "smaller than" in this context less than half the area of the WGS84 ellipsoid?

**Comment by Fil** [comment]
The text that was [commited](https://github.com/d3/d3-geo/commit/586c231272f5eed267a8830d0b08c1b107a0b119#diff-04c6e90faac2675aa89e2176d2eec7d8R15) just mentions that “Spherical polygons can be larger than a hemisphere” and links to an example where the polygon covers the Oceans https://bl.ocks.org/mbostock/6713736

In my understanding the winding order does not change with the polygon's size: if you project a small circle centered on [0,0] with a radius _r_ going from 0 to 180°, there is no discontinuity at 90° when _r_ crosses that threshold: the circle continues to grow in area and radius (though its circumference reverses). Of course it depends from which point of reference you define the "winding order" 👿

If we go to that ocean example, an analysis of the topojson shows that it is composed of two features:
- 0 = Caspian Sea
- 1 = the rest of the ocean(s)

Feature 1 is the most interesting: it is a Polygon (i.e. a list of rings); its first ring follows the coast of main Antarctic land in the reverse winding order, and thus defines an *enormous* surface covering almost all the Earth. Next rings are (lots of) holes, which correspond to the land polygons, in reverse winding order. (Which explains why the Caspian sea, being "a hole in a hole", is shown as an independent feature.)

I hope this clarifies rather that muddies the issue :)

I feel this situation with the two geojson defs is really painful. Maybe we'll need to add another set of convenience tools (similar to geoStitch), that would reverse geometries, maybe based on their area. There are sometimes geojson in the wild where the winding order changes from one polygon to the next… there is also in [a dark corner of d3-geo-projection](https://github.com/d3/d3-geo-projection/blob/c4b7b6d03150aab25444d67654f7562c80159fb1/src/project/index.js#L114) a test on the area to decide which way to look ;-)

**Comment by mbostock** [comment]
Yes, Philippe.

Perhaps we should avoid the terms “clockwise” and “counterclockwise” because these terms don’t make sense in the context of spherical polygons: depending on the point of reference, a spherical polygon is *both* clockwise *and* counterclockwise. You can talk about clockwise-ness after projecting to the plane, but even in that case a spherical polygon can become multiple planar polygons after clipping, and our definition of a spherical polygon should likewise be spherical.

Perhaps the definition should be something like “the area that is bounded by the polygon is to the right of the boundary” (per [ST_ForceRHR](https://postgis.net/docs/ST_ForceRHR.html)). So, if you walk the boundary of the polygon (in the defined order) along the surface of the sphere, the face of the polygon will be to your right.

Of course, it is tempting to call this the “right-hand rule” (as PostGIS does), but GeoJSON says it follows the right-hand rule with the opposite interpretation (for planar polygons).

**Comment by mbostock** [comment]
(Clockwise-ness is also confusing because in WGS84 +*y* points North (typically up) and in Canvas/SVG +*y* points down.)

**Comment by urschrei** [comment]
Ah, so that means you _can't_ re-orient an RFC 7946 planar polygon based on area (because both windings are valid in `d3-geo`, with effects as Philippe and you have explained)? To use them as spherical polygons, the only safe thing to do is to assume that the first `coordinates` ring is the exterior and orient it CW, and the subsequent rings are interior, orienting them CCW, before giving them to `d3-geo`?

**Comment by listenzcc** [comment]
In geometric,  a spherical polygon is both clockwise and counterclockwise.
And in practice, we can not make sure every datasets are providing coordinates in clockwise order.

So, is it possible to assume the polygons are of convex in prior?
Then the order of the coordinates can be detected automatically.
If it is provided as the 'wrong' order, just use the reverse of the list of points.

The demo code can be found at https://observablehq.com/@listenzcc/untitled.

**Comment by Fil** [comment]
See also https://github.com/mapbox/geojson-rewind

**Comment by listenzcc** [comment]
> See also https://github.com/mapbox/geojson-rewind

It’s a great solution. But there aren’t npm components in my web server. So It seems the rewind process has to be left to the front end in my case.

A tiny hope is that D3 geo package may consider to perform rewind optionally. Since it will inevitably unpack the json for each polygon to formulate the pathways.  It is convenient to do the job, isn’t it?

**Comment by jrus** [comment]
This description of 'right' vs. 'left' is very exactly the opposite of what I would expect, which would be: if you walk along the boundary in the specified direction, is the boundary to the right of the shape or the left of the shape.

I think clockwise / anticlockwise is clearer. You have to consider the direction compared to a point inside the shape though.

**Comment by Fil** [comment]
A better solution: https://observablehq.com/@fil/rewind

**Comment by matkoniecz** [comment]
As mentioned in #229 which got closed with message to post here:

It would be useful to have functionality with d3 to rewind right-hand geojson into left-hand geojson.
Or even ability to opt in in using RFC 7946 geojson, not d3 geojson.

Right now displaying RFC 7946 geojson requires importing extra library.

And while https://github.com/d3/d3-geo/pull/79#issuecomment-280935242 comments

> I’m disappointed that RFC 7946 standardizes the opposite winding order to D3, Shapefiles and PostGIS. And I don’t see an easy way for D3 to change its behavior, since it would break all existing (spherical) GeoJSON used by D3. I’d be interested to see an empirical analysis of GeoJSON use in practice to see what winding order is commonly used.

right-hand winding order seems to be clearly more and more standard to me (see for example QGIS which will not open them at all) and D3 seems to growing oddity. Especially as

> result is the same whether the exterior ring is clockwise or anticlockwise

appears to not be true in many cases

**Comment by mhkeller** [comment]
@Fil Are you planning on publishing your rewind utilities to npm for easier use outside of observable? It would be handy to be able to get any future updates, types and so on. Thanks for putting the examples together.

**Comment by Fil** [comment]
You mean https://observablehq.com/@fil/rewind ? I had no such plans, but this comes up often so it might be worth considering. If I wanted to do it I think I'd only expose `feature = rewind(feature);`, without the "simple=false" option, and none of the projection stuff. This could bring down the code to a few lines, and make the documentation quite straightforward, for 99% of the use cases.

**Comment by mhkeller** [comment]
Yep that's what I was referring to. It would also be useful to hear your thoughts about how this implementation differs from a couple others that I've seen out there, namely [@placemarkio/geojson-rewind](https://github.com/placemark/geojson-rewind#readme) (which is an updated version of the [mapbox library](https://github.com/mapbox/geojson-rewind) by the same author) and [@turf/rewind](https://www.npmjs.com/package/@turf/rewind/v/7.0.0). I did a deep dive into rewinding this past weekend after trying to get an export from PostGIS working and having to rewind seems like a fairly common problem but good documentation on a canonical way to solve it across platforms seems missing.

**Comment by Fil** [comment]
Both placemark and turf use a planar coordinate reference system. Since the determination of winding order depends on the area of the outer ring, some shapes risk being interpreted differently when they have a negative planar area but positive spherical area. (Not even mentioning other differences around the poles and antimeridian.)

This is not a pedantic/academic concern, but a practical one. If you use a spherical projection (say any of D3’s projections, except geoIdentity), you should rewind with a spherical function. If you use a planar-coordinates system such as leaflet and its derivatives, you should use a planar rewind function.

See the complete notebook with cases https://observablehq.com/@fil/spherical-winding-order-test-cases

**Comment by nshiab** [comment]
@mhkeller @Fil In case it could be useful, I added your implementation in [journalism](https://jsr.io/@nshiab/journalism/1.22.0/src/geo/rewind.ts) because I wanted to use it with [simple-data-analysis](https://github.com/nshiab/simple-data-analysis).
